Add flash loan tutorial (#465)
cowprotocol/docs: docs/cow-protocol/tutorials/cow-swap/flash-loans.mdx

---
sidebar_position: 5
---

# Flash loans

> The general setup for a flash loan requires:

Reviewer: Flashloans are surprisingly nuanced so instead of conveying "this is what you need" I'd suggest "this is the setup for this particular example". This tutorial is not meant to cover the entire solution space but rather one example of how things could work.

Author: Agree. I reframed the tutorial to focus entirely on the example.

* A flash loan lender (e.g. Aave).
* A use case for the funds (e.g. repaying a debt in Aave using a deposit as collateral).
* The necessary `ERC-20` approvals in the wallet: to CoW's Vault Relayer contract, as well as any additional approvals related to the use case (e.g. Aave pools).
* Some non-zero amount of the order's buy token in the wallet balance, to pass CoW Swap order validation.

## Example: repay debt with collateral using flash loans on Sepolia

We’ll walk through a concrete example of a debt repayment using flash loans on Sepolia.

There is an [example order](https://explorer.cow.fi/sepolia/orders/0x86ff28a37bf70c549edfe753d899d44307b1b4c3ae943c6f5bdfe271942fd4f135ed9a9d1122a1544e031cc92fcc7ea599e28d9c67dc66ab) for a particular case on Sepolia, including the [corresponding onchain transaction](https://sepolia.etherscan.io/tx/0x6f4fc9b450e08dab0ab2d935e7c7dcade9e44e330e484657e5b1fadb6c6e8851).
The example was executed using an [automated script](https://github.com/mrnaveira/cow-flashloan-tester) that places and signs a flash loan order on the Sepolia network.

### Test setup
The setup for the flash loan example is as follows:

> * A safe Wallet on address `0x35eD9A9D1122A1544e031Cc92fCC7eA599e28D9C`.

Reviewer: I think it's important to point out why we use a safe (that way a smart contract can initiate the withdrawal of the collateral token into the safe address).
Also I think it's worthwhile to point out that the test is using a `pre-sign` order but technically an `EIP-1271` signature would also work (it's actually strictly better for this use case).

* The Safe borrows 10,000 USDT on Aave using 400,000+ USDC as collateral.

The order itself intends to:
1. Request a flash loan to Aave of 5,000 USDT.
2. Use those 5,000 to partially repay the 10,000 USDT debt on Aave.

> 3. Withdraw 400,000 USDC from Aave.

Reviewer: Why can we get back the majority of our collateral if we only repay half of our debt? Was that actually the setup of your debt position?

Author: Health ratios in Aave on Sepolia end up being large (if there is enough liquidity to begin with) even with very small amounts of collateral, that's why we can withdraw most of it without being liquidated.

This was done on purpose to have as many tokens as possible available for the trade itself, as Uniswap ratios on Sepolia can be very extreme.

I updated the setup to specify the total amount of collateral (500,000) on the test setup, to avoid this confusion.

4. Perform a trade on CoW Swap, selling up to 400,000 USDC for 5,002.5 USDT (borrowed amount + [0.05% flash loan fee on Aaave](https://aave.com/docs/concepts/flash-loans)).
2. Use the 5,002.5 USDT from the trade to fully repay the flash loan (including the fee).

:::info

Liquidity and exchange ratios on the Sepolia network can be very inconsistent. To ensure successful execution of the order, we use a high amount of USDC (both as collateral and as the token being traded).

General considerations when testing flash loans on Sepolia and Aave:

> - [Token names in Aaave may not correspond](https://github.com/mrnaveira/cow-flashloan-tester?tab=readme-ov-file#2-select-aave-tokens) with tokens names in other dapps.

Reviewer: The names were fine it's just that we were dealing with different addresses, no?
```suggestion
- [Token addresses in Aave may not correspond](https://github.com/mrnaveira/cow-flashloan-tester?tab=readme-ov-file#2-select-aave-tokens) with tokens names in other dapps.
```

- [Token ratios](https://github.com/mrnaveira/cow-flashloan-tester?tab=readme-ov-file#4-determine-token-amounts-needed-to-trade-and-deposit) may need to be balanced, and trading amounts adjusted accordingly.
- Tokens in Aave can be [obtained from the faucet](https://github.com/mrnaveira/cow-flashloan-tester?tab=readme-ov-file#5-get-tokens-from-the-faucet).
- Check [withdrawal and borrowing availability](https://github.com/mrnaveira/cow-flashloan-tester?tab=readme-ov-file#7-check-aaves-withdrawalborrow-availability) of the tokens on Aave.

> - Try a different RPC in case of transaction submission issues.

Reviewer: This part seems a bit odd to point out since this applies to every web3 use case. 😅

:::

### Order construction

```json5
{
  "from": "Safe address",
  "sellToken": "USDC address", // collateral token (unlocked by repaying the debt)
  "buyToken": "USDT address", // originally borrowed token that was now advanced by the flash loan
  "buyAmount": 5002500000, // appData.flashloan.amount + 0.05% flashloan fee
  "receiver": "settlementContract", // this is for repaying the flash loan
  "validTo": "now + 5m", // managing risk can be done by having a short validity
  "kind": "buy", // buy exactly the flash loaned amount and keep the surplus in the collateral token
  "partiallyFillable": false, // if an order is partially fillable, then it is not ensured the debt will be paid
  "appData": {
    "hooks": {
      "pre": [
         // First: Repay the debt
         {
            "target": "0x123...321",
            "value": "0",
            "callData": "signedRepayCall"
         },
         // Then: Withdraw the collateral
         {
            "target": "0x123...321",
            "value": "0",
            "callData": "signedWithdrawCall"
         }
      ]
    },
    "flashloan": {
      "token": "USDT",
      "amount": 5000000000 // 5,000 USDT in atoms
    }
  }
}
```

> Check the [flash loan tester script](https://github.com/mrnaveira/cow-flashloan-tester) for more details on how this order is constructed.

Reviewer: Seems a bit weird to have a dedicated tutorial page and then defer to a repo. Why not copy the respective code parts into the docs itself and explain each step a bit more?
It's not a great example but I was thinking about sth like [this](https://fasterthanli.me/articles/catching-up-with-async-rust) for example.

Author: I reworked the tutorial to expand all step as suggested

### Setting the pre-signature of the order

After the order is placed, its pre-signature must be submitted to the CoW Settlement Contract via a transaction:
* Function [`setPresignature`](https://docs.cow.fi/cow-protocol/reference/contracts/core/settlement#setpresignature).
* Arguments: `orderId` and `true`.

The [flash loan tester script](https://github.com/mrnaveira/cow-flashloan-tester) already sends this transaction automatically to be executed by the Safe. You can see an example of this transaction [here](https://sepolia.etherscan.io/tx/0x2de6e4cfcf61942f73e2db254a6b917fc52ec95da4d296f5a998be4c1b910c94).

After the pre-signature transaction is confirmed, the order's status will be `Open`, and it will be eligible for execution within the CoW Protocol.

### Order execution

Once an order is placed within the CoW Protocol, it enters an auction batch. When a solution is found, the following steps occur:

> 1. The winning solver calls the flash loan `IFlashLoanRouter` contract.

Reviewer: I wonder if we could make this a bit more visual using a graphic like [here](https://app.blocksec.com/explorer/tx/sepolia/0x753b65e09447f1fb783b5aabbb4c1e35cf4311fb08a1d9695130e587a4f6812d).
We could color and number the individual token movements and explain what's happening at each step.

Author: IMO it's better to leave it for a follow-up PR

2. Use the 5,002.5 USDT from the trade to fully repay the flash loan (including the fee).

:::info

Exchange ratios on the Sepolia network can be very inconsistent. To ensure successful execution of the order, a large amount of USDC is withdrawn from Aave and used in the trade.

General considerations when testing flash loans on Sepolia and Aave:
- [Token addresses in Aave may not correspond](https://github.com/mrnaveira/cow-flashloan-tester?tab=readme-ov-file#2-select-aave-tokens) with tokens names in other dapps.
- [Token ratios](https://github.com/mrnaveira/cow-flashloan-tester?tab=readme-ov-file#4-determine-token-amounts-needed-to-trade-and-deposit) may need to be balanced, and trading amounts adjusted accordingly.
- Tokens in Aave can be [obtained from the faucet](https://github.com/mrnaveira/cow-flashloan-tester?tab=readme-ov-file#5-get-tokens-from-the-faucet).
- Check [withdrawal and borrowing availability](https://github.com/mrnaveira/cow-flashloan-tester?tab=readme-ov-file#7-check-aaves-withdrawalborrow-availability) of the tokens on Aave.

In a production network (e.g. mainnet), these considerations generally won’t apply.

:::

## Aave repay pre-hook

> We need to define a [pre-hook](/cow-protocol/reference/core/intents/hooks) to repay the 5,000 USDT debt on Aave.

Reviewer: Is this link getting rendered correctly when you build the docs?

Author: Yes, it renders correctly

> We need to define a [pre-hook](/cow-protocol/reference/core/intents/hooks) to repay the 5,000 USDT debt on Aave.

Reviewer: Being a bit more explicit with our context would be good.
i.e. we need to repay the debt so that the Aave pool allows us to withdraw the collateral from it.

As part of the settlement, we want to call the [`repay` function on Aave's pool](https://aave.com/docs/developers/smart-contracts/pool#write-methods-repay):
```solidity
function repay(
    address asset,
    uint256 amount,
    uint256 interestRateMode,
    address onBehalfOf
) public virtual override returns (uint256)
```

The following Typescript code generates the corresponding function data using [`viem`](https://viem.sh/):
```typescript
const repayTxData = encodeFunctionData({
    abi, // ABI definition for the repay function
    functionName: 'repay',
    args: [
      '0xaA8E23Fb1079EA71e0a56F48a2aA51851D8433D0', // asset: USDT
      '5000000000', // amount: 5,000 USDT (USDT has 6 decimals)
      2, // interestRateMode: 2 for variable
      '0x35eD9A9D1122A1544e031Cc92fCC7eA599e28D9C', // onBehalfOf: Safe address
    ]
  });
```

> The `repay` call must be executed during the order settlement process. To enable this, the pre-hook defines a transaction that will be sent to the Safe wallet. The Safe then executes the actual repay call to the Aave pool.

Reviewer: More specifically `repay` and `withdraw` absolutely have to be executed as a pre-hook because the settlement contract requires the user to have the collateral tokens before it pulls in the sell tokens of all settled orders.

The transaction can be built using the [Safe Protocol Kit](https://docs.safe.global/sdk/protocol-kit) as shown below:
```typescript
  const repaySafeTxData: SafeTransactionDataPartial = {
    to: '0x6Ae43d3271ff6888e7Fc43Fd7321a503ff738951', // Aave pool address in Sepolia
    value: '0',
    data: repayTxData, // created previously
    operation: OperationType.Call,
    nonce, // current Safe nonce + 1
  };

  const safeTransaction = await safe.createTransaction({ transactions: [repaySafeTxData] });
  const signedSafeTransaction = await safe.signTransaction(safeTransaction, SigningMethod.ETH_SIGN);
  const encodedSafeTransaction = await safe.getEncodedTransaction(signedSafeTransaction);
```

:::note

> For this example, the `nonce` should be offset by 1 from the current Safe's nonce, as this transaction will be executed in the future.

Reviewer: Seems unnecessary given that you provide all the context in the next sentence.
```suggestion
For this example, the `nonce` should be offset by 1 from the current Safe's nonce.
```

> This is needed to avoid using the same nonce twice, as we will send another transaction for the pre-signature in following sections.

Reviewer: Phrasing is slightly weird. Maybe this gives slightly more context.
```suggestion
This is needed to avoid using the same nonce twice, as we will send another transaction to emit the pre-signature for our order which is needed to make it executable in the first place.
```

:::

## Aave withdraw pre-hook

> We need to define another pre-hook to withdraw 400,000 USDC from the Aave deposit, needed for the trade.

Reviewer: Also slightly more context here:
We need to withdraw the collateral in a pre-hook because our safe needs to have the sell tokens needed for the order. Otherwise the settlement contract would not be able to pull the funds from the safe and the settlement would revert.

---
sidebar_position: 5
---

# Flash loans

We’ll walk through a concrete example of a debt repayment using flash loans on Sepolia.

There is an [example order](https://explorer.cow.fi/sepolia/orders/0x86ff28a37bf70c549edfe753d899d44307b1b4c3ae943c6f5bdfe271942fd4f135ed9a9d1122a1544e031cc92fcc7ea599e28d9c67dc66ab) for this particular case on Sepolia, including the [corresponding on-chain transaction](https://sepolia.etherscan.io/tx/0x6f4fc9b450e08dab0ab2d935e7c7dcade9e44e330e484657e5b1fadb6c6e8851).

> The example was executed using an [automated script](https://github.com/mrnaveira/cow-flashloan-tester) that places and signs a flash loan order on the Sepolia network. The following sections of the tutorial include code snippets from this tool.

Reviewer 2: This is pointing to personal GH repos, we shouldn't depend on repos controlled by individuals. Please consider moving this repo into COW.

Author: I moved the script to https://github.com/cowprotocol/cow-flashloan-tester

## Setup
The setup for the flash loan example is as follows:

> * A Safe wallet at address 0x35eD9A9D1122A1544e031Cc92fCC7eA599e28D9C. We use a smart contract wallet so that smart contracts can initiate the withdrawal of the collateral token.

Reviewer 2: nit: maybe `0x35eD9A9D1122A1544e031Cc92fCC7eA599e28D9C`, what do you think? 🤔

Reviewer 2: this sentence "We use a smart contract wallet so that smart contracts can initiate the withdrawal of the collateral token." could point out to the explanation I wrote (already in the docs) about why it has to be a contract.

* The Safe borrows 10,000 USDT on Aave using 500,000 USDC as collateral.
* Aave is used as the flash loan lender.

> * The necessary `ERC-20` approvals are set in the wallet — to CoW's Vault Relayer contract and to the relevant Aave pool contracts.

Reviewer 2: Can you please expand more here, which approvals and of which tokens (both , or just one depending on the operation, etc).

> * Some non-zero amount of the order's buy token in the Safe's balance, to pass CoW Swap order validation.

Reviewer 3: I guess most of the readers would look at this sentence and say "what, why?", but I don't have a suggestion how to improve it, maybe adding a bit more context here (but that would bloat the core idea around flashloans we are trying to explain here)

Author: Makes sense. I added just a tiny bit of context (_spam protection measure_).

            gasLimit: "1000000"
          },
          {
            target: config.SAFE_ADDRESS,
            value: "0",
            callData: withdrawTx, // withdraw pre-hook data
            gasLimit: "1000000"
          }
        ],
        "post": []
      },
      signer: config.SAFE_ADDRESS,
    },
  };
```

To programmatically create and submit a CoW Swap order, we can use [cow-sdk](https://github.com/cowprotocol/cow-sdk):
```typescript
  // Setup CoW SDK
  const sdk = new TradingSdk({
    chainId: SupportedChainId.SEPOLIA,
    signer: new VoidSigner(config.SAFE_ADDRESS, new JsonRpcProvider(config.RPC_URL)),
    appCode: 'Our flashloan example',
  });

  // Define trade parameters
  const parameters: TradeParameters = {
    env: 'staging', // Required for Sepolia
    kind: OrderKind.BUY,
    sellToken: USDC_ADDRESS,
    sellTokenDecimals: 6, // USDC has 6 decimals
    buyToken: USDT_ADDRESS,
    buyTokenDecimals: 6, // USDT has 6 decimals as well
    amount: '5002500000', // 5,002.5 USDT (borrowed amount + 0.05% Aaave flash loan fee)
    // receiver is always the settlement contract because the driver takes
    // funds from the settlement contract to pay back the loan

>     receiver: '0x9008D19f58AAbD9eD0D60971565AA8510560ab41', // cow settlement contract address in sepolia

Reviewer 3: `on Sepolia`

      orderId,
      true,
    ]
  });

  const safeTransactionData: SafeTransactionDataPartial = {
    to: "0x9008D19f58AAbD9eD0D60971565AA8510560ab41", // CoW settlement contract on Sepolia
    value: '0',
    data: txData,
    operation: OperationType.Call,
  }

  const safeTransaction = await safe.createTransaction({ transactions: [safeTransactionData] });
  const signedSafeTransaction = await safe.signTransaction(safeTransaction, SigningMethod.ETH_SIGN);

  const transactionResult = await safe.executeTransaction(signedSafeTransaction);
  console.log('setPreSignature transaction hash: ' + transactionResult.hash);
```

After the pre-signature transaction is confirmed, the order's status will be `Open`, making it eligible for execution within the CoW Protocol.

:::note
This example uses a pre-signature, but a valid EIP-1271 contract signature could also be used.
With an EIP-1271 signature, the `setPreSignature` transaction would not be necessary.
:::

## Order execution

Once an order is placed within the CoW Protocol, it enters an auction batch. When a solution is found, the following steps occur:

1. The winning solver calls the flash loan `IFlashLoanRouter` contract.

> 2. The 5,000 USDT gets transferred to the flash loan `IFlashLoanRouter` contract.

Reviewer 2: You can link `IFlashLoanRouter` from the contract page?
